automap/src/comm_layer/pmp.rs

```diff
@@ -396,6 +448,139 @@ impl ThreadGuts {
+                    "{}\n{}",
+                    err_msg,
+                    PrettyHex::hex_dump(&buffer)
+                );
+                Err(AutomapError::ProtocolError(err_msg))
+            }
+        }
+    }
+
+    fn handle_announcement(&self, public_ip: Ipv4Addr, mapping_config_opt: &Option<MappingConfig>) {
+        match mapping_config_opt {
+            Some(mapping_config) => {
+                let mut packet = PmpPacket {
+                    opcode: Opcode::MapTcp,
+                    direction: Direction::Request,
+                    ..Default::default()
+                };
+                let opcode_data = MapOpcodeData {
+                    epoch_opt: None,
+                    internal_port: mapping_config.hole_port,
+                    external_port: mapping_config.hole_port,
+                    lifetime: mapping_config.next_lifetime_secs(),
+                };
+                packet.opcode_data = Box::new(opcode_data);
+                debug!(
+                    &self.logger,
+                    "Sending mapping request to {} and waiting for response", self.router_addr
+                );
+                match PmpTransactor::transact(
+                    &self.factories_arc,
+                    self.router_addr,
+                    &packet,
+                    PMP_READ_TIMEOUT_MS,
+                    &self.logger,
+                ) {
+                    Ok(response) => match response.result_code_opt {
+                        Some(ResultCode::Success) => {
+                            debug!(
+                                &self.logger,
+                                "Remapping successful; triggering change handler"
+                            );
+                            self.change_handler.as_ref()(AutomapChange::NewIp(IpAddr::V4(
+                                public_ip,
+                            )));
+                        }
+                        Some(result_code) => {
+                            let err_msg = format!(
+                                "Remapping after IP change failed; Node is useless: {:?}",
+                                result_code
+                            );
+                            error!(&self.logger, "{}\n{:?}", err_msg, packet);
+                            let automap_error = if result_code.is_permanent() {
+                                AutomapError::PermanentMappingError(err_msg)
+                            } else {
+                                AutomapError::TemporaryMappingError(err_msg)
+                            };
+                            self.change_handler.as_ref()(AutomapChange::Error(automap_error));
+                        }
+                        None => {
+                            let err_msg = "Remapping after IP change failed; Node is useless: Received request when expecting response".to_string();
+                            error!(&self.logger, "{}\n{:?}", err_msg, packet);
```

Comment:
Maybe some kind of more formal continuation between the err message and the packet? If you print the variable `packet` it will look like it is the broken packet that came back, but what you really print is our request that stimulated the wrong response.

Maybe you want to print the response instead or with the original request together...

Reply:
Yep, I wanted the response instead. Good catch.